### PR TITLE
[AVR] Fix parsing & emitting relative jumps

### PR DESCRIPTION
**Pull request by Patryk27**
Ever since 6859685a87ad093d60c8bed60b116143c0a684c7 (or, precisely, 84428dafc0941e3a31303fa1b286835ab2b8e234) relative jumps emitted by the AVR codegen are off by two bytes - this pull request fixes it.

## Abstract

As compared to absolute jumps, relative jumps - such as rjmp, rcall or brsh - have an implied `pc+2` behavior; that is, `jmp 100` is `pc = 100`, but `rjmp 100` gets understood as `pc = pc + 100 + 2`.

This is not reflected in the AVR codegen:

https://github.com/llvm/llvm-project/blob/f95026dbf66e353128a3a3d7b55f3e52d5985535/llvm/lib/Target/AVR/MCTargetDesc/AVRAsmBackend.cpp#L89

... which always emits relative jumps that are two bytes too far - or rather it _would_ emit such jumps if not for this check:

https://github.com/llvm/llvm-project/blob/f95026dbf66e353128a3a3d7b55f3e52d5985535/llvm/lib/Target/AVR/MCTargetDesc/AVRAsmBackend.cpp#L517

... which causes most of the relative jumps to be actually resolved late, by the linker, which applies the offsetting logic on its own, hiding the issue within LLVM.

[Some time ago](https://github.com/llvm/llvm-project/commit/697a162fa63df328ec9ca334636c5e85390b2bf0) we've had a similar "jumps are off" problem that got solved by touching `shouldForceRelocation()`, but I think that has worked only by accident. It's exploited the fact that absolute vs relative jumps in the parsed assembly can be distinguished through a "side channel" check relying on the existence of labels (i.e. absolute jumps happen to named labels, but relative jumps are anonymous, so to say). This was an alright idea back then, but it got broken by 6859685a87ad093d60c8bed60b116143c0a684c7.

I propose a different approach:
- when emitting relative jumps, offset them by `-2` (well, `-1`, strictly speaking, because those instructions rely on right-shifted offset),
- when parsing relative jumps, treat `.` as `+2` and read `rjmp .+1234` as `rjmp (1234 + 2)`.

This approach seems to be sound and now we generate the same assembly as avr-gcc, which can be confirmed with:

```cpp
// avr-gcc test.c -O3 && avr-objdump -d a.out

int main() {
    asm(
"      foo:\n\t"
"        rjmp  .+2\n\t"
"        rjmp  .-2\n\t"
"        rjmp  foo\n\t"
"        rjmp  .+8\n\t"
"        rjmp  end\n\t"
"        rjmp  .+0\n\t"
"      end:\n\t"
"        rjmp .-4\n\t"
"        rjmp .-6\n\t"
"      x:\n\t"
"        rjmp x\n\t"
"        .short 0xc00f\n\t"
);
}
```

avr-gcc is also how I got the opcodes for all new tests like `inst-brbc.s`, so we should be good.

cc @aykevl

### CODE REVIEW

**Comment by aykevl** [comment]
I'll make a backport request.

**Comment by aykevl** [comment]
Done: https://github.com/llvm/llvm-project/issues/106728

**Comment by Patryk27** [comment]
Ouch, I've had no idea there's an extra LLD test (`lld/test/ELF/avr-reloc.s`) - lemme take a look at it.

**Comment by aykevl** [comment]
Oops, I ran `ninja check-llvm` but not `ninja check-lld` or something

**Comment by nico** [comment]
Looks like this broke check-lld: http://45.33.8.238/macm1/91316/summary.html

Please take a look and revert for now if it takes a while to fix

**Comment by Patryk27** [comment]
I'm preparing a fix right now, should be a simple thing.

**Comment by aykevl** [comment]
@Patryk27 thanks for fixing this!

File: llvm/test/MC/AVR/inst-brhc.s [hunk 1]
```diff
@@ -0,0 +1,24 @@
+; RUN: llvm-mc -triple avr -show-encoding < %s | FileCheck %s
+;
+; RUN: llvm-mc -filetype=obj -triple avr < %s \
+; RUN:     | llvm-objdump -d - \
+; RUN:     | FileCheck --check-prefix=INST %s
+
+foo:
+  brhc .+12
```
Comment (MaskRay) [inline llvm/test/MC/AVR/inst-brhc.s:8]: Having so many instructions could be a maintenance problem.

For other targets we typically pack tests of the same instruction format.

llvm/test/MC/M68k/ and llvm/test/MC/LoongArch/ are good examples.

Comment (Patryk27) [reply llvm/test/MC/AVR/inst-brhc.s:8]: Ok, thanks - I'll take a look!

**Pull request merged**